src/util/std_expr.h

```diff
@@ -3031,17 +3023,16 @@ inline void validate_expr(const address_of_exprt &value)
 
 /*! \brief Boolean negation
 */
-class not_exprt:public exprt
+class not_exprt:public unary_exprt
 {
 public:
-  explicit not_exprt(const exprt &op):exprt(ID_not, bool_typet())
+  explicit not_exprt(const exprt &op):
+    unary_exprt(ID_not, op, bool_typet())
```

Comment:
The use of `bool_typet()` should be unnecessary here, and in fact would be wrong if `op.type()!=bool_typet()`.

Reply:
ID_not is really only meant to be used with boolean operands (this could be a precondition).
Note we've also got ID_bitnot.

Reply:
Omitting the `bool_typet()` argument will yield additional sharing (as op.type() would just get copied). I agree that the precondition ought to be added as well.

Reply:
Yes, agreed on the sharing opportunity. Will give that a go, together with precondition.